ch_tools/chadmin/internal/table.py

```diff
@@ -249,6 +260,136 @@ def delete_table(
     )
 
 
+def _is_valid_uuid(uuid_str):
+    try:
+        val = uuid.UUID(uuid_str)
+    except ValueError:
+        return False
+    return str(val) == uuid_str
+
+
+def _get_uuid_table(table_metadata_path: str) -> str:
+    table_uuid = ""
+
+    with open(table_metadata_path, "r", encoding="utf-8") as metadata_file:
+        for line in metadata_file:
+            if "UUID" in line:
+                parts = line.split()
+                try:
+                    uuid_index = parts.index("UUID")
+                    # check size
+                    table_uuid = parts[uuid_index + 1].strip("'")
+                except ValueError:
+                    raise RuntimeError(
+                        f"Failed parse UUID from '{table_metadata_path}'"
+                    )
+                break
+        else:
+            raise RuntimeError(f"No UUID in '{table_metadata_path}'")
+
+    assert _is_valid_uuid(table_uuid)
+
+    return table_uuid
+
+
+def check_table_dettached(ctx, database_name, table_name):
+    query = """
+        SELECT
+            1
+        FROM system.tables
+        WHERE database = '{{ database_name }}'
+          AND table = '{{ table_name }}'
+    """
+
+    response = execute_query(
+        ctx,
+        query,
+        database_name=database_name,
+        table_name=table_name,
+        format_="JSON",
+    )["data"]
+
+    if 0 != len(response):
+        raise RuntimeError(
+            f"Table '{database_name}'.'{table_name}' is attach. Use delete without --detach flag."
+        )
+
+
+def _has_object_storage(ctx: Context) -> bool:
+    query = """
+        SELECT 1 FROM system.disks WHERE name = 'object_storage'
+    """
+    response = execute_query(
+        ctx,
+        query,
+        format_="JSON",
+    )["data"]
+
+    if 0 == len(response):
+        logging.info("No object storage.")
+        return False
+
+    logging.info("There is an object storage.")
+    return True
+
+
+def _remove_table_data_from_object_storage(local_metadata_table_path):
+    table_uuid = _get_uuid_table(local_metadata_table_path)
+
+    object_storage_table_data_path = "store" + "/" + table_uuid[:3] + "/" + table_uuid
+
+    logging.info(
+        "Table has UUID {}, path in S3: {}.", table_uuid, object_storage_table_data_path
+    )
+
+    disk = "object_storage"
+    disk_config_path = make_ch_disks_config(disk)
+
+    code, stderr = remove_from_ch_disk(
+        disk=disk,
+        path=object_storage_table_data_path,
+        disk_config_path=disk_config_path,
+    )
+    if code:
+        raise RuntimeError(
+            f"clickhouse-disks remove command has failed: retcode {code}, stderr: {stderr.decode()}"
+        )
+
+
+def delete_detached_table(ctx, database_name, table_name):
+    check_table_dettached(ctx, database_name, table_name)
+
+    local_metadata_table_path = (
+        CLICKHOUSE_METADATA_PATH + "/" + database_name + "/" + table_name + ".sql"
```

Comment:
ClickHouse escapes database and table names when constructing paths on file system. So this won't work for arbitrary names.